### PR TITLE
rgw: prevent spurious/lost notifications in the index completion thread

### PR DESCRIPTION
**Pull request by yuvalif**
this was happening when async completions happened during reshard.
more information about testing:
https://gist.github.com/yuvalif/d526c0a3a4c5b245b9e951a6c5a10517

to trace possible issues, dpp was added to the completion thread.
this should allow finding unhandled completions due to reshards. for example:
https://0x0.st/oKQ1.log

Signed-off-by: Yuval Lifshitz <ylifshit@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by adamemerson — APPROVED** [review]

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -832,120 +833,66 @@ struct complete_op_data {
...
-  {
-    std::lock_guard l{completions_lock};
-    completions.swap(comps);
-  }
+class RGWIndexCompletionManager {
+  RGWRados* const store;
+  const int num_shards;
+  ceph::containers::tiny_vector<ceph::mutex> locks;
+  std::vector<set<complete_op_data*>> completions;
+  std::vector<complete_op_data*> async_completions;
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply src/rgw/rgw_rados.cc:841]: > the `async_` piece is a bit misleading here? it's `completions` that were submitted asynchronously with `aio_operate()`. the completions in this vector failed with ERR_BUSY_RESHARDING and need to be retried synchronously. maybe call this `retry_completions`?
> 

sure, will rename

> what was the motivation for collapsing `RGWIndexCompletionThread` into `RGWIndexCompletionManager`? i think it's better to keep the `Thread` class and encapsulate its state/locking there. that way we don't mix `completions/locks` and `async_completions/async_completions_lock` together in one place

there is no thread class. it is just a function called from the manager class inside its own thread.
both the async_completion_lock is used inside functions called from different thread contexts, os it belongs to the manager class

**Comment by yuvalif** [comment]
* when testing with 2 zones (1 rgw in each zone), and uploading objects only to primary i do not see any difference between the zones
* when testing with 2 zones (1 rgw in each zone), and uploading objects only to the same bucket in both zones, there is no difference between the objects, however, the 2ndary zone sync status is "behind". see:
https://paste.sh/aGzWOiJ8#UyGOOVvdILeFknWgd-nioBXd

File: src/rgw/rgw_rados.h [hunk 2]
```diff
@@ -1398,13 +1398,13 @@ class RGWRados
...
-  int cls_obj_complete_add(BucketShard& bs, const rgw_obj& obj, std::string& tag, int64_t pool, uint64_t epoch, rgw_bucket_dir_entry& ent,
+  int cls_obj_complete_add(const DoutPrefixProvider *dpp, BucketShard& bs, const rgw_obj& obj, std::string& tag, int64_t pool, uint64_t epoch, rgw_bucket_dir_entry& ent,
                            RGWObjCategory category, std::list<rgw_obj_index_key> *remove_objs, uint16_t bilog_flags, rgw_zone_set *zones_trace = nullptr);
-  int cls_obj_complete_del(BucketShard& bs, std::string& tag, int64_t pool, uint64_t epoch, rgw_obj& obj,
+  int cls_obj_complete_del(const DoutPrefixProvider *dpp, BucketShard& bs, std::string& tag, int64_t pool, uint64_t epoch, rgw_obj& obj,
                            ceph::real_time& removed_mtime, std::list<rgw_obj_index_key> *remove_objs, uint16_t bilog_flags, rgw_zone_set *zones_trace = nullptr);
-  int cls_obj_complete_cancel(BucketShard& bs, std::string& tag, rgw_obj& obj,
+  int cls_obj_complete_cancel(const DoutPrefixProvider *dpp, BucketShard& bs, std::string& tag, rgw_obj& obj,
```
Comment (cbodley) [inline src/rgw/rgw_rados.h:1407]: this adds a `dpp` to lots of functions, apparently for `RGWIndexCompletionManager::create_completion()`, but the new `dpp` argument is unused there. can we safely remove these now? that would simplify backports

Comment (yuvalif) [reply src/rgw/rgw_rados.h:1407]: ok. removing dpp. as it is too expensive to cache it inside the callback args

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
merged, thanks @yuvalif!

**Pull request closed without merging**